api/src/game_state_server.py

```diff
@@ -0,0 +1,157 @@
+from dataclasses import dataclass
+
+UNIT_SIZE = 50
+
+
+@dataclass
+class Token:
+
+    id: str
+    icon_id: str
+    start_x: int
+    start_y: int
+    start_z: int
+    end_x: int
+    end_y: int
+    end_z: int
+
+    def to_dict(self):
+
+        return {
+            'id': self.id,
+            'icon_id': self.icon_id,
+            'start_x': self.start_x,
+            'start_y': self.start_y,
+            'start_z': self.start_z,
+            'end_x': self.end_x,
+            'end_y': self.end_y,
+            'end_z': self.end_z,
+        }
+
+
+class RoomData:
+
+    def __init__(self, room_id, initial_connection=None):
+
+        self.room_id = room_id
+        self.game_state = {}
+        self.id_to_positions = {}
+        self.positions_to_ids = {}
+        self.clients = set()
+        if initial_connection:
+            self.clients.add(initial_connection)
+
+
+class GameStateServer:
+
+    def __init__(self):
+
+        self._rooms = {}
+
+    def new_connection_request(self, client, room_id):
+
+        if self._rooms.get(room_id, False):
+            self._rooms[room_id].clients.add(client)
+            return self.get_state(room_id)
+        else:
+            self._rooms[room_id] = RoomData(room_id, initial_connection=client)
+            return
+
+    def connection_dropped(self, client, room_id):
+
+        if self._rooms.get(room_id, False):
+            self._rooms[room_id].clients.remove(client)
+
+    def process_updates(self, messages, room_id):
+
+        if not (self._rooms.get(room_id, False) and self._rooms[room_id].clients):
+            print('Unable to process update from new connection')
```

Comment:
In this case, I like throwing an exception and the websocketmanager can catch it and send an error back to the client